crates/librqbit_core/src/magnet.rs

```diff
@@ -2,41 +2,61 @@ use std::str::FromStr;
 
 use anyhow::Context;
 
-use crate::id20::Id20;
+use crate::hash_id::{Id20, Id32};
+
 
 /// A parsed magnet link.
 pub struct Magnet {
-    pub info_hash: Id20,
+    id20: Option<Id20>,
+    id32: Option<Id32>,
     pub trackers: Vec<String>,
 }
 
 impl Magnet {
+    pub fn as_id20(&self) -> Option<Id20> {
+        self.id20
+    }
+
+    pub fn as_id32(&self) -> Option<Id32> {
+        self.id32
+    }
+
     /// Parse a magnet link.
     pub fn parse(url: &str) -> anyhow::Result<Magnet> {
         let url = url::Url::parse(url).context("magnet link must be a valid URL")?;
         if url.scheme() != "magnet" {
             anyhow::bail!("expected scheme magnet");
         }
-        let mut info_hash: Option<Id20> = None;
+        let mut info_hash_found = false;
+        let mut id20: Option<Id20> = None;
+        let mut id32: Option<Id32> = None;
         let mut trackers = Vec::<String>::new();
         for (key, value) in url.query_pairs() {
             match key.as_ref() {
-                "xt" => match value.as_ref().strip_prefix("urn:btih:") {
-                    Some(infohash) => {
-                        info_hash.replace(Id20::from_str(infohash)?);
+                "xt" => {
+                    if let Some(ih) = value.as_ref().strip_prefix("urn:btih:") {
+                        let i = Id20::from_str(ih)?;
+                        id20.replace(i);
+                        info_hash_found = true;
+                    } else if let Some(ih) = value.as_ref().strip_prefix("urn:btmh:1220") {
+                        let i = Id32::from_str(ih)?;
+                        id32.replace(i);
+                        info_hash_found = true;
+                    } else {
+                        anyhow::bail!("expected xt to start with btih or btmh");
                     }
-                    None => anyhow::bail!("expected xt to start with urn:btih:"),
                 },
                 "tr" => trackers.push(value.into()),
                 _ => {}
             }
         }
-        match info_hash {
-            Some(info_hash) => Ok(Magnet {
-                info_hash,
+        match info_hash_found {
```

Comment:
nit: you don't need a separate variable to track this, could have just check if at least one of id20 and id32 is set